File: solutions/javascript/run-length-encoding/1/run-length-encoding.js
const getCode = (count, letter) => {
  return count === 1 ? letter : `${count}${letter}`;
};

export const encode = (text) => {
  if (!text) return text;

  let encoded = '';
  let letter  = text[0];
  let count   = 1;

  for (let i = 1; i < text.length; i++) {
    const current = text.charAt(i);

    if (current === letter) {
      count++;
    }
    else {
      encoded += getCode(count, letter);

      letter = current;
      count  = 1;
    }
  }

  encoded += getCode(count, letter);
  return encoded;
};

export const decode = (text) => {
  let decoded = '';
  let digits  = '';

  for (let i = 0; i < text.length; i++) {
    const current = text.charAt(i);

    let digit = current.match(/\d/);

    if (digit) {
      digits += digit;
    }
    else {
      const count = digits ? +digits : 1;

      for (let j = 0; j < count; j++) {
        decoded += current;
      }

      digits = '';
    }
  }

  return decoded;
};

File: solutions/lua/run-length-encoding/1/run-length-encoding.lua
local get_code = function(count, c)
  if count == 1 then return c end
  return string.format("%d%s", count, c)
end

return {
  encode = function(s)
    local coded   = ""
    local current = s:sub(1, 1)
    local count   = 1

    for i = 2, #s do
      local c = s:sub(i, i)
      if c == current then
        count = count + 1
      else
        coded   = coded .. get_code(count, current)
        current = c
        count   = 1
      end
    end

    return coded .. get_code(count, current)
  end,
  decode = function(s)
    local plain  = ""
    local digits = ""
    local count  = 0

    for i = 1, #s do
      local c = s:sub(i, i)
      if c:match("%d") then
        digits = digits .. c
      else
        if digits == "" then count = 1 else count = tonumber(digits) end
        plain  = plain .. c:rep(count)
        digits = ""
      end
    end

    return plain
  end
}
